### PR TITLE
Use statefulset for placement service for HA mode

### PR DESCRIPTION
**Pull request by youngbupark**
# Description

> NOTE: test will be failed until #2411 is merged.

Use statefulset for placement service for HA mode:
1. Change from Deployment to statefulset for placement deployment
   - Placement will need to know the addresses of each peer node to establish raft RPC connections and save the snapshot and logs in the persistent volume (PV). With PV, when placement restarts, it reads the log entries and snapshots to recover the state. So we need to deploy placement as statefulset service.
2. Use headless service for placement
3. Change vnode size from 1000 to 100 - 1000 occupies too much memory. In the test, 100 Actor PODs are deployed which results in OOMKill for the container. We need to document it well about how to adjust the parameter for user's environment.
4. Change the placement service port from 80 to 50005 because headless service cannot forward port. [breaking-change]
5. Update helm chart readme

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #2376

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by youngbupark** [comment]
@yaron2 Switching to statefulset will break upgrade scenario of placement. The option to not break the scenario is that we can change `dapr-placement` service name to the different name e.g. `dapr-placement-srv` or `placement`

1. Run helm upgrade
2. Restart all applications to use new stateful placement service
3. Delete old `dapr-placement` service

If we do not like to change the name of service, we can ask users to uninstall and install the services, but the existing production users would not like it. 

What do you think? if you have better idea, please let me know.

File: charts/dapr/charts/dapr_placement/templates/dapr_placement_service.yaml [hunk 1]
```diff
@@ -2,10 +2,17 @@ kind: Service
...
+    app: dapr-placement
 spec:
   selector:
     app: dapr-placement
+  # placement must be able to resolve pod address to join initial cluster peers
+  # before POD is ready
+  publishNotReadyAddresses: true
   ports:
-  - protocol: TCP
-    port: {{ .Values.ports.port }}
```
Comment (youngbupark) [inline charts/dapr/charts/dapr_placement/templates/dapr_placement_service.yaml:10]: headless service doesn't support `targetPort` so we need to change the placement port from 80 to 50005

**Comment by youngbupark** [comment]
**NOTE: test will be failed until #2411 is merged.**

File: Makefile [hunk 2]
```diff
@@ -17,6 +17,8 @@ GIT_VERSION = $(shell git describe --always --abbrev=7 --dirty)
 CGO         ?= 0
 BINARIES    ?= daprd placement operator injector sentry
 HA_MODE     ?= false
+# Force in-memory log for placement
+FORCE_INMEM ?=true
```
Comment (artursouza) [inline Makefile:21]: nit: space

File: Makefile [hunk 3]
```diff
@@ -196,8 +199,11 @@ upload-helmchart:
 docker-deploy-k8s: check-docker-env check-arch
 	$(info Deploying ${DAPR_REGISTRY}/${RELEASE_NAME}:${DAPR_TAG} to the current K8S context...)
 	$(HELM) install \
-		$(RELEASE_NAME) --namespace=$(DAPR_NAMESPACE) --wait --timeout 5m0s\
-		--set global.ha.enabled=$(HA_MODE) --set-string global.tag=$(DAPR_TAG)-$(TARGET_OS)-$(TARGET_ARCH) --set-string global.registry=$(DAPR_REGISTRY) --set global.logAsJson=true --set global.daprControlPlaneOs=$(TARGET_OS) --set global.daprControlPlaneArch=$(TARGET_ARCH) $(HELM_CHART_DIR)
+		$(RELEASE_NAME) --namespace=$(DAPR_NAMESPACE) --wait --timeout 5m0s\ 
+		--set global.ha.enabled=$(HA_MODE) --set-string global.tag=$(DAPR_TAG)-$(TARGET_OS)-$(TARGET_ARCH)\ 
+		--set-string global.registry=$(DAPR_REGISTRY) --set global.logAsJson=true\ 
+		--set global.daprControlPlaneOs=$(TARGET_OS) --set global.daprControlPlaneArch=$(TARGET_ARCH)\ 
+		--set dapr_placement.cluster.forceInMemoryLog=$(FORCE_INMEM) $(HELM_CHART_DIR)
```
Comment (tcnghia) [inline Makefile:206]:
```suggestion
		$(RELEASE_NAME) --namespace=$(DAPR_NAMESPACE) --wait --timeout 5m0s \
		--set global.ha.enabled=$(HA_MODE) --set-string global.tag=$(DAPR_TAG)-$(TARGET_OS)-$(TARGET_ARCH) \
		--set-string global.registry=$(DAPR_REGISTRY) --set global.logAsJson=true \
		--set global.daprControlPlaneOs=$(TARGET_OS) --set global.daprControlPlaneArch=$(TARGET_ARCH) \
		--set dapr_placement.cluster.forceInMemoryLog=$(FORCE_INMEM) $(HELM_CHART_DIR)
```

**Review by tcnghia — COMMENTED** [review]
Can you add some doc & PR description to highlight why StatefulSet is used instead of Deployment?

File: Makefile (hunk 2 above)
Comment (youngbupark) [reply Makefile]: done

**Comment by youngbupark** [comment]
> Can you add some doc & PR description to highlight why StatefulSet is used instead of Deployment?

I will add doc separately. TL;DR; Raft needs to save log entries and snapshot in the persistent volume for the recovery and make consensus among the peers. So StatefulSet + persistent volume enables this scenarios. Also, there should be the option to use inmemory instead of persistent volume.

**Comment by youngbupark** [comment]
DNS resolving issue on windows - https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#dns-limitations

(Thanks @mukundansundar )

File: pkg/injector/pod_patch.go [hunk 4]
```diff
@@ -114,7 +114,7 @@ func (i *injector) getPodPatchOperations(ar *v1.AdmissionReview,
 	}
 
 	// Keep DNS resolution outside of getSidecarContainer for unit testing.
-	placementAddress := fmt.Sprintf("%s:80", getKubernetesDNS(placementService, namespace))
+	placementAddress := fmt.Sprintf("%s:50005", getKubernetesDNS(placementService, namespace))
```
Comment (mukundansundar) [inline pkg/injector/pod_patch.go:117]: why is the port hardcoded here but override-able in helm chart?

Comment (youngbupark) [reply pkg/injector/pod_patch.go:117]: value in helm-chart is used in the multiple places and we will not document it in readme.md

**Review by artursouza — APPROVED** [review]

**Review by mukundansundar — APPROVED** [review]

**Pull request merged**